File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-hook-shebang.ts
#!/usr/bin/env bun
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Require hooks to have correct shebang.
 */
export const requireHookShebangRule = (ctx: RuleContext): HarnessCheckRule => ({
	applies(manifest: HarnessManifest): boolean {
		const section = manifest.requireHookShebang;
		if (typeof section !== "object" || section === null) {
			return false;
		}
		const enabled = (section as { enabled?: unknown }).enabled;
		return (
			enabled !== false &&
			ctx.readStringArray(
				ctx.readJsonObject((section as Record<string, unknown>).parameters)
					.hooks,
			).length > 0
		);
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		const parameters = ctx.readJsonObject(
			ctx.readJsonObject(manifest.requireHookShebang).parameters,
		);
		const hooks = ctx.readStringArray(parameters.hooks);
		const expectedShebang =
			typeof parameters.expectedShebang === "string"
				? parameters.expectedShebang
				: "#!/usr/bin/env sh";
		return hooks.flatMap((hook) => {
			return ctx.isFile(hook) && ctx.firstLine(hook) !== expectedShebang
				? [
						{
							severity: ctx.severityOf(manifest, "requireHookShebang"),
							category: "requireHookShebang",
							message: `${hook} must start with ${expectedShebang}`,
						},
					]
				: [];
		});
	},
});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_hook_shebang.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require hook shebang rule."""

from collections.abc import Iterable

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import first_line, severity_for

class RequireHookShebangRule(HarnessCheckRule):
    """Validate requireHookShebang check."""

    category = "requireHookShebang"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate requireHookShebang check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        hooks = params.get("hooks", [])
        if not isinstance(hooks, list):
            return []
        expected_shebang = params.get("expectedShebang", "#!/usr/bin/env sh")
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "{hook} must start with {expectedShebang}")
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                template.format(hook=hook, expectedShebang=expected_shebang),
            )
            for hook in hooks
            if isinstance(hook, str) and first_line(root / hook) != expected_shebang
        ]
